### ISSUE TITLE
using rootfs with credential isolation causes permission denied if the target rootfs folder isn't world executable

### ISSUE THREAD

**Issue by ocanty** [nginx/unit#464]
Example application:

```
{
  "type": "php",
  "root": f"/",
  "isolation": {
      "namespaces": {
          "cgroup": True,
          "credential": True,
          "mount": True,
          "network": True,
          "pid": True,
          "uname": True
      },
      "uidmap": [{
          "host": 1000,
          "container": 0,
          "size": 1
      }],
      "gidmap": [{
          "host": 1000,
          "container": 0,
          "size":  1
      }],
      "rootfs": f"/home/test/website"
  },
  "user": f"root",
  "group": f"root"
}
```

`/home/test` is owned by `test` (id 1000) and has permissions `751`
`/home/test/website` is owned by `test` and has permissions `751`

This works fine and the application is setup correctly

However, if you want to prevent other users from listing the contents of `/home/test` and setting it to `750`
This will cause the application to not start, triggering an error similar to  `2020/08/14 05:36:28 [alert] 1559#1559 error bind mounting rootfs (13: Permission denied)`

Since the issue was being triggered by world permissions, I tried to add uid/gidmaps to map nobody/nogroup in the container to the host nobody/nogroup (65534->65534) to see if that would work but it did not

My unitd is executing with the command line `--user root --group root` as mentioned in the rootfs blogpost

**Comment by artemkonev** [comment]
Hi, @ocanty! Have you tried setting the runtime options ``--user`` or ``--group`` to ``test``?
I'm trying to reproduce the issue and it seems that the router process needs the directory permissions before the mount actually happens. Let's see what @i4ki has to say, though.

**Comment by i4ki** [comment]
Hi @ocanty 

> My unitd is executing with the command line --user root --group root (this might be incorrect if I misinterpreted the recent blog post about rootfs)

Yes, this is incorrect. The only requirement is running Unit _as root_ (eg.: sudo unitd).

About using a 750 (or 700) _rootfs_ directory: Yes, you got a corner case we didn't test before. Let me explain the issue first:

When you use `"credential": true`, the process is placed in a new USER namespace and has all capabilities set in that namespace but it has *no capabilities* in the parent namespace. It means we need to chroot(2)/pivot_root(2) before applying the uidmap/gidmap into the new namespace, and then the file system's checks are failing because the user-id and group-id are unknown to the file system driver at this point. 

By giving an *x* access bit to *others* is enough (x means access for directories) to make the permission check pass because `mount/chroot/chdir` only test this bit.

In other words, we have a chicken-and-egg kind of issue here, where we can't apply uidmap because we need to call `chroot` and `mount` but also we can't `mount` or `chdir` into rootfs if we don't apply (in this case of _protected_ directory).

The solution maybe would be record the `x` permission bit of _others_ in the directory and `chmod` this bit if needed, then do the mount and chroot calls, and restore it later.

A workaround for you now can be: Do a `chmod 751 $rootfs` (or 701), apply the config, and then restore it to 750 afterward.

I'll discuss with the team some ways for improving this.

Thank you for reporting it.

**Comment by i4ki** [comment]
@artemkonev The PHP bind mounts for the extension directory wasn't released yet.

This is an issue for every application type if the rootfs directory has no permission for other users.

**Comment by tippexs** [comment]
@ac000 and @alejandro-colomar as we are working on some isolation testings and cases, i would like to keep this issue open and add it to our isolation work list. Happy to talk about it on Monday. 👍🏼
